File: backend-csharp/Controllers/SessionController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Picture2Text.Api.DTOs.Responses;
using Picture2Text.Api.Services;
using System.Security.Claims;

namespace Picture2Text.Api.Controllers;

/// <summary>
/// 會話管理控制器
/// </summary>
[Authorize]
[ApiController]
[Route("api/[controller]")]
public class SessionController : ControllerBase
{
    private readonly SessionService _sessionService;
    private readonly LoginHistoryService _loginHistoryService;

    public SessionController(SessionService sessionService, LoginHistoryService loginHistoryService)
    {
        _sessionService = sessionService;
        _loginHistoryService = loginHistoryService;
    }

    /// <summary>
    /// 取得當前使用者的所有活躍會話
    /// </summary>
    /// <returns>會話列表</returns>
    [HttpGet("active")]
    [ProducesResponseType(typeof(SessionListResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    public async Task<ActionResult<SessionListResponse>> GetActiveSessions()
    {
        var userId = int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)!.Value);
        var currentSessionId = User.FindFirst("sessionId")?.Value;

        var sessions = await _sessionService.GetUserActiveSessionsAsync(userId);

        var sessionInfos = sessions.Select(s => new SessionInfo
        {
            SessionId = s.SessionId,
            DeviceName = s.DeviceName,
            IpAddress = s.IpAddress,
            LoginAt = s.LoginAt,
            LastActivityAt = s.LastActivityAt,
            IsCurrent = s.SessionId == currentSessionId,
            ExpiresAt = s.ExpiresAt
        }).ToList();

        return Ok(new SessionListResponse
        {
            Code = 200,
            Message = "獲取成功",
            Data = sessionInfos
        });
    }

    /// <summary>
    /// 結束指定的會話
    /// </summary>
    /// <param name="sessionId">會話 ID</param>
    /// <returns>操作結果</returns>
    [HttpDelete("{sessionId}")]
    [ProducesResponseType(typeof(ApiResponse<object>), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ApiResponse<object>), StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    public async Task<ActionResult<ApiResponse<object>>> EndSession(string sessionId)
    {
        var userId = int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)!.Value);
        
        // 驗證會話屬於當前使用者
        var sessions = await _sessionService.GetUserActiveSessionsAsync(userId);
        var session = sessions.FirstOrDefault(s => s.SessionId == sessionId);

        if (session == null)
        {
            return NotFound(new ApiResponse<object>
            {
                Code = 404,
                Message = "會話不存在或已結束",
                Data = null
            });
        }

        var success = await _sessionService.EndSessionAsync(sessionId);

        if (!success)
        {
            return NotFound(new ApiResponse<object>
            {
                Code = 404,
                Message = "會話不存在",
                Data = null
            });
        }

        return Ok(new ApiResponse<object>
        {
            Code = 200,
            Message = "會話已結束",
            Data = null
        });
    }

    /// <summary>
    /// 取得登入歷史記錄
    /// </summary>
    /// <param name="limit">返回記錄數量（預設 50）</param>
    /// <returns>登入歷史列表</returns>
    [HttpGet("history")]
    [ProducesResponseType(typeof(LoginHistoryResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    public async Task<ActionResult<LoginHistoryResponse>> GetLoginHistory([FromQuery] int limit = 50)
    {
        var userId = int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)!.Value);

        var histories = await _loginHistoryService.GetUserLoginHistoryAsync(userId, limit);

        var historyInfos = histories.Select(h => new LoginHistoryInfo
        {
            Id = h.Id,
            IsSuccess = h.IsSuccess,
            FailureReason = h.FailureReason,
            IpAddress = h.IpAddress,
            DeviceInfo = h.DeviceInfo,
            AttemptedAt = h.AttemptedAt,
            Location = h.Location
        }).ToList();

        return Ok(new LoginHistoryResponse
        {
            Code = 200,
            Message = "獲取成功",
            Data = historyInfos
        });
    }
}

File: backend-nestjs/src/controllers/session.controller.ts
import {
  Controller,
  Get,
  Delete,
  Param,
  Query,
  Res,
  HttpStatus,
  UseGuards,
} from '@nestjs/common';
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiBearerAuth,
  ApiParam,
  ApiQuery,
} from '@nestjs/swagger';
import { Response } from 'express';
import { SessionService, LoginHistoryService, RefreshTokenService } from '../services';
import { ApiResponse as ApiRes, SessionListResponse, LoginHistoryResponse } from '../dto/response';
import { CurrentUser, CurrentUserData } from '../decorators/current-user.decorator';
import { JwtAuthGuard } from '../guards';

@ApiTags('Session')
@Controller('api/session')
@UseGuards(JwtAuthGuard)
@ApiBearerAuth()
export class SessionController {
  constructor(
    private readonly sessionService: SessionService,
    private readonly loginHistoryService: LoginHistoryService,
    private readonly refreshTokenService: RefreshTokenService,
  ) {}

  /**
   * 取得使用者的活躍 Session 列表
   */
  @Get('active')
  @ApiOperation({ summary: '取得活躍 Session', description: '取得當前使用者的所有活躍 Session' })
  @ApiResponse({ status: 200, description: 'Session 列表', type: SessionListResponse })
  @ApiResponse({ status: 401, description: '未授權' })
  async getActiveSessions(
    @CurrentUser() user: CurrentUserData,
    @Res() res: Response,
  ) {
    const sessions = await this.sessionService.getUserActiveSessionsAsync(
      user.userId,
      user.sessionId,
    );

    return res.status(HttpStatus.OK).json(ApiRes.success(sessions));
  }

  /**
   * 結束指定的 Session
   */
  @Delete(':sessionId')
  @ApiOperation({ summary: '結束 Session', description: '結束指定的 Session（登出該裝置）' })
  @ApiParam({ name: 'sessionId', description: 'Session ID' })
  @ApiResponse({ status: 200, description: '結束成功' })
  @ApiResponse({ status: 401, description: '未授權' })
  @ApiResponse({ status: 404, description: 'Session 不存在' })
  async endSession(
    @Param('sessionId') sessionId: string,
    @CurrentUser() user: CurrentUserData,
    @Res() res: Response,
  ) {
    // 取得 Session 以獲取 RefreshTokenId
    const session = await this.sessionService.getSessionByIdAsync(sessionId);

    if (!session || session.UserId !== user.userId) {
      return res.status(HttpStatus.NOT_FOUND).json(ApiRes.notFound('Session 不存在或無權限'));
    }

    // 結束 Session
    const success = await this.sessionService.endSessionAsync(sessionId, user.userId);

    if (!success) {
      return res.status(HttpStatus.NOT_FOUND).json(ApiRes.notFound('Session 不存在'));
    }

    // 撤銷對應的 Refresh Token
    if (session.RefreshTokenId) {
      const refreshToken = await this.refreshTokenService.getByTokenIdAsync(
        session.RefreshToken?.TokenId || '',
      );
      if (refreshToken) {
        await this.refreshTokenService.revokeRefreshTokenAsync(refreshToken.Token);
      }
    }

    return res.status(HttpStatus.OK).json(ApiRes.success(null, 'Session 已結束'));
  }

  /**
   * 取得登入歷史記錄
   */
  @Get('history')
  @ApiOperation({ summary: '取得登入歷史', description: '取得當前使用者的登入歷史記錄' })
  @ApiQuery({ name: 'limit', required: false, description: '回傳筆數（預設 20）' })
  @ApiResponse({ status: 200, description: '登入歷史', type: LoginHistoryResponse })
  @ApiResponse({ status: 401, description: '未授權' })
  async getLoginHistory(
    @Query('limit') limit: string | undefined,
    @CurrentUser() user: CurrentUserData,
    @Res() res: Response,
  ) {
    const historyLimit = limit ? parseInt(limit, 10) : 20;
    const history = await this.loginHistoryService.getUserLoginHistoryAsync(
      user.userId,
      historyLimit,
    );

    return res.status(HttpStatus.OK).json(ApiRes.success(history));
  }
}
